src/merkle/smt/mod.rs

```diff
@@ -743,19 +779,57 @@ impl<const DEPTH: u8, K: Serializable, V: Serializable> Serializable for Mutatio
     fn write_into<W: ByteWriter>(&self, target: &mut W) {
         target.write(self.old_root);
         target.write(self.new_root);
-        self.node_mutations.write_into(target);
-        self.new_pairs.write_into(target);
+
+        let inner_removals: Vec<_> = self
+            .node_mutations
+            .iter()
+            .filter(|(_, value)| matches!(value, NodeMutation::Removal))
+            .map(|(key, _)| key)
+            .collect();
+        let inner_additions: Vec<_> = self
+            .node_mutations
+            .iter()
+            .filter_map(|(key, value)| match value {
+                NodeMutation::Addition(node) => Some((key, node)),
+                _ => None,
+            })
+            .collect();
+
+        target.write_usize(inner_removals.len());
+        target.write_many(inner_removals);
+
+        target.write_usize(inner_additions.len());
+        target.write_many(inner_additions);
+
+        target.write_usize(self.new_pairs.len());
+        target.write_many(&self.new_pairs);
     }
 }
 
-impl<const DEPTH: u8, K: Deserializable + Ord, V: Deserializable> Deserializable
+impl<const DEPTH: u8, K: Deserializable + KeyConstraints, V: Deserializable> Deserializable
     for MutationSet<DEPTH, K, V>
 {
     fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
         let old_root = source.read()?;
         let new_root = source.read()?;
-        let node_mutations = source.read()?;
-        let new_pairs = source.read()?;
+
+        let num_removals = source.read_usize()?;
+        let inner_removals: Vec<NodeIndex> = source.read_many(num_removals)?;
```

Comment:
Similar comment to the one above: I think we should be able to read the vectors directly.